File: 560.和为-k-的子数组.py
#
# @lc app=leetcode.cn id=560 lang=python3
#
# [560] 和为 K 的子数组
#

# @lc code=start
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        # 子数组：连续，又要求和，考虑前缀和
        dict = {} # 记录次数
        sum_val = res = 0
        dict[sum_val] = 1
        for item in nums:
            sum_val += item
            if (sum_val - k) in dict: # 前缀和，表示中间一段为k
                res += dict[sum_val - k] # 有几种，就有几种子数组，起点不同，终点相同
            dict[sum_val] = dict.get(sum_val, 0) + 1 # 出现该值的字数
        return res
        

# @lc code=end

File: 560.和为-k-的子数组.java
/*
 * @lc app=leetcode.cn id=560 lang=java
 *
 * [560] 和为 K 的子数组
 */

// @lc code=start
import java.util.HashMap;
class Solution {
    public int subarraySum(int[] nums, int k) {
        //前缀和
        int res = 0;
        int n = nums.length;
        int[] prefix = new int[n+1]; //默认为0
        HashMap<Integer, Integer> map = new HashMap<>();
        map.put(0,1);
        for(int i=1;i<n+1;i++){
            prefix[i] = prefix[i-1] + nums[i-1];
            int target = prefix[i] - k;
            res += map.getOrDefault(target, 0);
            map.put(prefix[i], map.getOrDefault(prefix[i], 0)+1);
        }
        return res;
    }
}
// @lc code=end

